File: Backend/app/models/sub_category.rb
class SubCategory < ApplicationRecord
  belongs_to :category
  has_many :products, dependent: :destroy

  has_one_attached :imagen
  
  validates :nombre, presence: true
  
  before_validation :generate_slug, on: [:create, :update]
  

  def imagen_url
    return nil unless imagen.attached?
    
    bucket = ENV['AWS_BUCKET']
    region = ENV['AWS_REGION']
    key = imagen.key
    
    "https://#{bucket}.s3.#{region}.amazonaws.com/#{key}"
  end

  def generate_slug
    base_slug = nombre.parameterize
    slug_candidate = base_slug
    count = 2

    while SubCategory.exists?(slug: slug_candidate)
      slug_candidate = "#{base_slug}-#{count}"
      count += 1
    end

    self.slug = slug_candidate
  end 
end

File: Frontend/ale_beauty_art_app/lib/models/subcategory.dart
class SubCategory {
  final int id;
  final String nombre;
  final String imagen; // imagen_url expuesta por el backend
  final String slug;
  final Category? category;

  SubCategory({
    required this.id,
    required this.nombre,
    this.imagen = '',
    this.slug = '',
    this.category,
  });

  factory SubCategory.fromJson(Map<String, dynamic> json) {
    return SubCategory(
      id: json['id'],
      nombre: json['nombre'],
      imagen: json['imagen_url'] ?? '',
      slug: json['slug'] ?? '',
      category:
          json['category'] != null ? Category.fromJson(json['category']) : null,
    );
  }
}

class Category {
  final int id;
  final String nombreCategoria;

  Category({
    required this.id,
    required this.nombreCategoria,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'],
      nombreCategoria: json['nombre_categoria'],
    );
  }
}
